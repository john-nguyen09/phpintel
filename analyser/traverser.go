package analyser

import (
	"github.com/john-nguyen09/go-phpparser/phrase"
)

func Traverse(node phrase.AstNode, analyser *Analyser) {
	if p, ok := node.(*phrase.Phrase); ok {
		analyser.Preorder(p)

		for _, child := range p.Children {
			Traverse(child, analyser)
		}

		analyser.Postorder(p)
	}
}
