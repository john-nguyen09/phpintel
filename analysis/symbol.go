package analysis

import (
	"github.com/john-nguyen09/phpintel/analysis/storage"
	"github.com/john-nguyen09/phpintel/internal/lsp/protocol"
	sitter "github.com/smacker/go-tree-sitter"
)

// VisibilityModifierValue is a value of visibility modifier (public, protected, private)
type VisibilityModifierValue int

const (
	// Public indicates public visibility
	Public VisibilityModifierValue = iota
	// Protected indicates protected visibility
	Protected = iota
	// Private indicates private visibility
	Private = iota
)

func (v VisibilityModifierValue) ToString() string {
	if v == Public {
		return "public"
	}
	if v == Private {
		return "private"
	}
	if v == Protected {
		return "protected"
	}
	return ""
}

// ClassModifierValue is a value of class modifier (abstract, final)
type ClassModifierValue int

const (
	// NoClassModifier indicates there is no class modifier
	NoClassModifier ClassModifierValue = iota
	// Abstract indicates there is an abstract keyword
	Abstract = iota
	// Final indicates there is a final keyword
	Final = iota
)

// Symbol is a symbol
type Symbol interface {
	GetLocation() protocol.Location
}

// NameIndexable indicates a symbol is name indexable, i.e. have completion
type NameIndexable interface {
	GetIndexableName() string
	GetIndexCollection() string
}

type HasParams interface {
	GetParams() []*Parameter
	GetDescription() string
	GetNameLabel() string
}

type HasScope interface {
	GetScope() string
	IsScopeSymbol() bool
}

type serialisable interface {
	GetCollection() string
	GetKey() string
	Serialise(*storage.Encoder)
}

type HasParamsResolvable interface {
	ResolveToHasParams(ctx ResolveContext) []HasParams
}

func transformQualifiedName(n *sitter.Node, document *Document) TypeString {
	return NewTypeString(document.GetNodeText(n))
}
