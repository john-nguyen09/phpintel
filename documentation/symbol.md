# Symbol
## Definition
- Is internal structure for indexer to understand
- Can have multiple types:
    - Symbol
    - TransformSymbol
    - CollectionSymbol
    - TokenSymbol
- Interfaces:
    - Consumer
    - Reference
    - ScopeMember
- Is indexed when File consumes