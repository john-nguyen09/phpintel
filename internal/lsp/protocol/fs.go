package protocol

import (
	"context"
	"io/ioutil"

	"github.com/john-nguyen09/phpintel/internal/jsonrpc2"
	"github.com/john-nguyen09/phpintel/util"
	"github.com/karrick/godirwalk"
)

// FS is an interface for reading files
type FS interface {
	ConvertToURI(uri string) string
	ReadFile(ctx context.Context, uri string) ([]byte, error)
	ListFiles(ctx context.Context, base string) ([]TextDocumentIdentifier, error)
}

// URIFS is a LSP file reader which relies on xcontentProvider and xfilesProvider
// client capabilities
type URIFS struct {
	conn *jsonrpc2.Conn
}

var _ FS = (*URIFS)(nil)

// NewLSPFS creates an URIFS instance from the connection
func NewLSPFS(conn *jsonrpc2.Conn) *URIFS {
	return &URIFS{
		conn: conn,
	}
}

// ConvertToURI does nothing and return the path because it's already URI
func (f *URIFS) ConvertToURI(path string) string {
	return path
}

// ReadFile uses textDocument/xcontent request to read content of the uri
func (f *URIFS) ReadFile(ctx context.Context, uri string) ([]byte, error) {
	var textDocument TextDocumentItem
	err := f.conn.Call(ctx, "textDocument/xcontent", &ContentParams{
		TextDocument: TextDocumentIdentifier{
			URI: uri,
		},
	}, &textDocument)
	if err != nil {
		return nil, err
	}
	return []byte(textDocument.Text), nil
}

// ListFiles uses workspace/xfiles request to list files under a base
func (f *URIFS) ListFiles(ctx context.Context, base string) (docs []TextDocumentIdentifier, err error) {
	err = f.conn.Call(ctx, "workspace/xfiles", FilesParams{Base: base}, &docs)
	return
}

// FileFS uses file path
type FileFS struct{}

// NewFileFS creates a FileFS instance
func NewFileFS() *FileFS {
	return &FileFS{}
}

var _ FS = (*FileFS)(nil)

// ConvertToURI converts path to URI
func (f *FileFS) ConvertToURI(path string) string {
	return util.PathToURI(path)
}

// ReadFile reads the file from uri, given that the uri is the file path
// rather than actually the URI
func (f *FileFS) ReadFile(ctx context.Context, uri string) ([]byte, error) {
	path, err := util.URIToPath(uri)
	if err != nil {
		return nil, err
	}
	return ioutil.ReadFile(path)
}

// ListFiles lists files from the base
func (f *FileFS) ListFiles(ctx context.Context, base string) ([]TextDocumentIdentifier, error) {
	var results []TextDocumentIdentifier
	godirwalk.Walk(base, &godirwalk.Options{
		Callback: func(path string, de *godirwalk.Dirent) error {
			if !de.IsDir() {
				results = append(results, TextDocumentIdentifier{
					URI: path,
				})
			}
			return nil
		},
		Unsorted: true,
	})
	return results, nil
}
