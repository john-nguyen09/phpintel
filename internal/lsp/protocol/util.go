package protocol

import (
	"strconv"
	"strings"
)

// IsInRange compare whether a position is within a range
func IsInRange(position Position, theRange Range) int {
	start := theRange.Start
	end := theRange.End

	if position.Line < start.Line || (position.Line == start.Line && position.Character < start.Character) {
		return -1
	}
	if position.Line > end.Line || (position.Line == end.Line && position.Character >= end.Character) {
		if ComparePos(start, end) == 0 {
			return -1
		}
		return 1
	}
	return 0
}

// ComparePos returns -1, 0 or 1 indicating whether a is before, equal to or after b
func ComparePos(a Position, b Position) int {
	if a.Line < b.Line {
		return -1
	}
	if a.Line == b.Line {
		if a.Character < b.Character {
			return -1
		}
		if a.Character == b.Character {
			return 0
		}
		if a.Character > b.Character {
			return 1
		}
	}
	return 1
}

// CompareRange compares 2 ranges
func CompareRange(a Range, b Range) int {
	if IsInRange(a.End, b) < 0 {
		return -1
	}
	if IsInRange(a.Start, b) > 0 {
		return 1
	}
	return 0
}

// PosFromString creates Position from string
func PosFromString(str string) Position {
	values := strings.Split(str, ":")
	line, err := strconv.Atoi(values[0])
	if err != nil {
		panic(err)
	}
	character, err := strconv.Atoi(values[1])
	if err != nil {
		panic(err)
	}
	return Position{
		Line:      line,
		Character: character,
	}
}

// RangeFromString creates Range from string
func RangeFromString(str string) Range {
	values := strings.Split(str, "-")
	start := PosFromString(values[0])
	end := PosFromString(values[1])
	return Range{
		Start: start,
		End:   end,
	}
}
