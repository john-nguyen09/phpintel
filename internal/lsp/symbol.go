package lsp

import (
	"context"

	"github.com/john-nguyen09/phpintel/analysis"
	"github.com/john-nguyen09/phpintel/internal/lsp/protocol"
)

func toDocumentSymbolRecursive(symbol analysis.Symbol) protocol.DocumentSymbol {
	docSymbol, _ := symbolToProtocolDocumentSymbol(symbol)
	if docSymbol.Kind == 0 {
		return docSymbol
	}
	if block, ok := symbol.(analysis.BlockSymbol); ok {
		for _, child := range block.GetChildren() {
			childDocSymbol := toDocumentSymbolRecursive(child)
			if childDocSymbol.Kind == 0 {
				continue
			}
			docSymbol.Children = append(docSymbol.Children, childDocSymbol)
		}
	}
	return docSymbol
}

func (s *Server) documentSymbol(ctx context.Context, params *protocol.DocumentSymbolParams) ([]protocol.DocumentSymbol, error) {
	symbols := []protocol.DocumentSymbol{}
	uri := params.TextDocument.URI
	store := s.store.getStore(params.TextDocument.URI)
	if store == nil {
		return symbols, nil
	}
	document := store.GetOrCreateDocument(uri)
	if document == nil {
		return symbols, nil
	}
	document.Load()

	results := []protocol.DocumentSymbol{}
	for _, child := range document.Children {
		docSymbol := toDocumentSymbolRecursive(child)
		if docSymbol.Kind == 0 {
			continue
		}
		results = append(results, docSymbol)
	}
	return results, nil
}

func symbolToProtocolDocumentSymbol(symbol analysis.Symbol) (protocol.DocumentSymbol, string) {
	switch v := symbol.(type) {
	case *analysis.Class:
		return protocol.DocumentSymbol{
			Kind:           protocol.Class,
			Name:           v.GetName(),
			Detail:         v.GetDescription(),
			Range:          v.GetLocation().Range,
			SelectionRange: v.GetLocation().Range,
		}, v.Name.GetFQN()
	case *analysis.Const:
		return protocol.DocumentSymbol{
			Kind:           protocol.Constant,
			Name:           v.GetName(),
			Detail:         v.GetDescription(),
			Range:          v.GetLocation().Range,
			SelectionRange: v.GetLocation().Range,
		}, ""
	case *analysis.ClassConst:
		return protocol.DocumentSymbol{
			Kind:           protocol.Constant,
			Name:           v.GetName(),
			Detail:         v.GetDescription(),
			Range:          v.GetLocation().Range,
			SelectionRange: v.GetLocation().Range,
		}, ""
	case *analysis.Define:
		return protocol.DocumentSymbol{
			Kind:           protocol.Constant,
			Name:           v.GetName(),
			Detail:         v.GetDescription(),
			Range:          v.GetLocation().Range,
			SelectionRange: v.GetLocation().Range,
		}, ""
	case *analysis.Function:
		return protocol.DocumentSymbol{
			Kind:           protocol.Function,
			Name:           v.GetName().GetOriginal(),
			Detail:         v.GetDescription(),
			Range:          v.GetLocation().Range,
			SelectionRange: v.GetLocation().Range,
		}, ""
	case *analysis.Interface:
		return protocol.DocumentSymbol{
			Kind:           protocol.Interface,
			Name:           v.Name.GetOriginal(),
			Detail:         v.GetDescription(),
			Range:          v.GetLocation().Range,
			SelectionRange: v.GetLocation().Range,
		}, v.Name.GetFQN()
	case *analysis.Method:
		return protocol.DocumentSymbol{
			Kind:           protocol.Method,
			Name:           v.Name,
			Detail:         v.GetDescription(),
			Range:          v.GetLocation().Range,
			SelectionRange: v.GetLocation().Range,
		}, ""
	case *analysis.Property:
		return protocol.DocumentSymbol{
			Kind:           protocol.Property,
			Name:           v.Name,
			Detail:         v.GetDescription(),
			Range:          v.GetLocation().Range,
			SelectionRange: v.GetLocation().Range,
		}, ""
	case *analysis.Trait:
		return protocol.DocumentSymbol{
			Kind:           protocol.Class,
			Name:           v.GetName(),
			Detail:         v.GetDescription(),
			Range:          v.GetLocation().Range,
			SelectionRange: v.GetLocation().Range,
		}, v.Name.GetFQN()
	}
	return protocol.DocumentSymbol{}, ""
}

func (s *Server) workspaceSymbol(ctx context.Context, params *protocol.WorkspaceSymbolParams) ([]protocol.SymbolInformation, error) {
	symbols := []protocol.SymbolInformation{}
	if params.Query != "" {
		opts := baseSearchOptions()
		for _, store := range s.store.stores {
			classes, _ := store.SearchClasses(params.Query, opts)
			for _, class := range classes {
				symbols = append(symbols, protocol.SymbolInformation{
					Kind:     protocol.Class,
					Name:     class.Name.GetFQN(),
					Location: class.GetLocation(),
				})
			}
			consts, _ := store.SearchConsts(params.Query, opts)
			for _, constant := range consts {
				symbols = append(symbols, protocol.SymbolInformation{
					Kind:     protocol.Constant,
					Name:     constant.Name.GetFQN(),
					Location: constant.GetLocation(),
				})
			}
			classConsts, _ := store.SearchClassConsts("", params.Query, opts)
			for _, classConst := range classConsts {
				symbols = append(symbols, protocol.SymbolInformation{
					Kind:          protocol.Constant,
					Name:          classConst.Name,
					Location:      classConst.GetLocation(),
					ContainerName: classConst.Scope.GetFQN(),
				})
			}
			defines, _ := store.SearchDefines(params.Query, opts)
			for _, define := range defines {
				symbols = append(symbols, protocol.SymbolInformation{
					Kind:     protocol.Constant,
					Name:     define.Name.GetFQN(),
					Location: define.GetLocation(),
				})
			}
			functions, _ := store.SearchFunctions(params.Query, opts)
			for _, function := range functions {
				symbols = append(symbols, protocol.SymbolInformation{
					Kind:     protocol.Function,
					Name:     function.Name.GetFQN(),
					Location: function.GetLocation(),
				})
			}
			interfaces, _ := store.SearchInterfaces(params.Query, opts)
			for _, theInterface := range interfaces {
				symbols = append(symbols, protocol.SymbolInformation{
					Kind:     protocol.Interface,
					Name:     theInterface.Name.GetFQN(),
					Location: theInterface.GetLocation(),
				})
			}
			methods, _ := store.SearchMethods("", params.Query, opts)
			for _, method := range methods {
				symbols = append(symbols, protocol.SymbolInformation{
					Kind:          protocol.Method,
					Name:          method.Name,
					Location:      method.GetLocation(),
					ContainerName: method.Scope.GetFQN(),
				})
			}
			properties, _ := store.SearchProperties("", params.Query, opts)
			for _, property := range properties {
				symbols = append(symbols, protocol.SymbolInformation{
					Kind:          protocol.Property,
					Name:          property.Name,
					Location:      property.GetLocation(),
					ContainerName: property.Scope.GetFQN(),
				})
			}
			traits, _ := store.SearchTraits(params.Query, opts)
			for _, trait := range traits {
				symbols = append(symbols, protocol.SymbolInformation{
					Kind:     protocol.Class,
					Name:     trait.Name.GetFQN(),
					Location: trait.GetLocation(),
				})
			}
		}
	}
	return symbols, nil
}
