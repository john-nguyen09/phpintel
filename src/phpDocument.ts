import { Phrase, Parser } from "../node_modules/php7parser";

export class PhpDocument {
    constructor(public uri: string, public text: string) { }

    getTree(): Phrase {
        return Parser.parse(this.text);
    }
}