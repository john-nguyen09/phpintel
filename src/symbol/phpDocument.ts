import { Phrase, Parser } from "php7parser";
import { Consumer, Symbol, needsNameResolve } from "./symbol";
import { NamespaceDefinition } from "./namespace/definition";
import { ImportTable } from "../type/importTable";
import { NamespaceUse } from "./namespace/Use";
import { nonenumerable } from "../util/decorator";
import { Class } from "./class/class";
import { Constant } from "./constant/constant";
import { Function } from "./function/function";
import { ClassConstant } from "./constant/classConstant";
import { Method } from "./function/method";
import { Property } from "./variable/property";
import { isReference, Reference } from "./reference";

export class PhpDocument extends Symbol implements Consumer {
    @nonenumerable
    public text: string;

    @nonenumerable
    private _uri: string;

    public modifiedTime: number = -1;
    public importTable: ImportTable;

    public classes: Class[] = [];
    public functions: Function[] = [];
    public constants: Constant[] = [];
    public classConstants: ClassConstant[] = [];
    public methods: Method[] = [];
    public properties: Property[] = [];
    public references: Reference[] = [];

    constructor(uri: string, text: string) {
        super();

        this._uri = uri;
        this.text = text;
        this.importTable = new ImportTable();
    }

    get uri(): string {
        return this._uri;
    }

    getTree(): Phrase {
        return Parser.parse(this.text);
    }

    getOffset(line: number, character: number): number {
        let lines = this.text.split('\n');
        let slice = lines.slice(0, line);

        return slice.map((line) => {
            return line.length;
        }).reduce((total, lineCount) => {
            return total + lineCount;
        }) + slice.length + character;
    }

    consume(other: Symbol): boolean {
        if (other instanceof NamespaceDefinition) {
            this.importTable.setNamespace(other.name);

            return true;
        } else if (other instanceof NamespaceUse) {
            for (let alias of other.aliasTable) {
                this.importTable.import(alias.fqn, alias.alias);
            }

            return true;
        }

        if (needsNameResolve(other)) {
            other.resolveName(this.importTable);
        }

        return true;
    }

    pushSymbol(symbol: Symbol | null): void {
        if (symbol === null) {
            return;
        }

        if (symbol instanceof Class) {
            this.classes.push(symbol);
        } else if (symbol instanceof Function) {
            this.functions.push(symbol);
        } else if (symbol instanceof Constant) {
            this.constants.push(symbol);
        } else if (symbol instanceof ClassConstant) {
            this.classConstants.push(symbol);
        } else if (symbol instanceof Method) {
            this.methods.push(symbol);
        } else if (symbol instanceof Property) {
            this.properties.push(symbol);
        }
        
        if (isReference(symbol)) {
            this.references.push(symbol);
        }
    }
}