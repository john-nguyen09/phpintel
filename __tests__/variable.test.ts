import "reflect-metadata";
import * as path from "path";
import * as fs from "fs";
import { Traverser } from "../src/traverser";
import { SymbolParser } from "../src/symbol/symbolParser";
import { pathToUri } from "../src/util/uri";
import { PhpDocument } from "../src/symbol/phpDocument";
import { Parser } from "php7parser";
import { dumpAstToDebug } from "../src/testHelper";

describe('variable', () => {
    it('simple variable', () => {
        let filePath = path.resolve(__dirname, '..', 'case', 'variable', 'simpleVariable.php');
        let treeTraverser = new Traverser();

        const fileUri = pathToUri(filePath);
        const fileContent = fs.readFileSync(filePath).toString();
        let symbolParser = new SymbolParser(new PhpDocument(
            fileUri,
            fileContent
        ));
        let parseTree = Parser.parse(fileContent);

        treeTraverser.traverse(parseTree, [
            symbolParser
        ]);

        dumpAstToDebug(path.join('variable', 'simpleVariable.ast.json'), parseTree);
        
        expect(symbolParser.getTree().toObject()).toMatchSnapshot();
    });
});