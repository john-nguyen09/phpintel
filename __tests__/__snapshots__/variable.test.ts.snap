// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`variable simple variable 1`] = `
PhpDocument {
  "classConstants": Array [],
  "classes": Array [],
  "constants": Array [],
  "functions": Array [],
  "importTable": ImportTable {
    "imports": Object {},
    "namespace": NamespaceName {
      "parts": Array [],
    },
  },
  "methods": Array [],
  "modifiedTime": -1,
  "properties": Array [],
  "references": Array [
    SimpleVariable {
      "name": "$varName",
      "type": TypeComposite {
        "_types": Array [
          TypeName {
            "name": "string",
          },
        ],
        "existingTypes": Object {
          "string": true,
        },
      },
    },
    SimpleVariable {
      "name": "$fieldName",
      "type": TypeComposite {
        "_types": Array [],
        "existingTypes": Object {},
      },
    },
  ],
  "scopeVarStack": Array [
    ScopeVar {
      "location": Object {
        "range": Object {
          "end": 39,
          "start": 0,
        },
        "uri": "case/variable/simpleVariable.php",
      },
      "variables": Object {
        "$varName": TypeComposite {
          "_types": Array [
            TypeName {
              "name": "string",
            },
          ],
          "existingTypes": Object {
            "string": true,
          },
        },
      },
    },
  ],
  "text": "<?php
$varName = 'field_' . $fieldName;",
  "uri": "case/variable/simpleVariable.php",
}
`;
